src/MonitorProfiler/MainProfiler/ExceptionTracker.cpp

```diff
@@ -0,0 +1,311 @@
+        - ExceptionSearchFilterLeave: Invoked when finishing the evaluation of the filter.
+      - ExceptionSearchCatcherFound: Invoked only if a matching filter is found.
+      - ExceptionSearchFunctionLeave: Invoked when finishing the inspection of the function for a matching handler.
+      If a matching handler was found, the exception catcher finding phase ends after the ExceptionSearchFunctionLeave callback.
+    - Phase 3: Unwinding.
+      The following are invoked for each frame starting at the leaf frame (the frame from which the exception is thrown):
+      - ExceptionUnwindFunctionEnter: Invoked when beginning to unwind a frame.
+      If a matching handler was found, the unwind phase ends after the invocation of
+      the ExceptionUnwindFunctionEnter callback for the frame that contains the matching handler.
+      - ExceptionUnwindFunctionLeave: Invoked when finishing the unwind of a frame.
+      If all of the frames are unwound to completion, the process is terminated.
+
+    Unhandled Exception Detection Algorithm
+
+    A subset of the aboved described callbacks is used to determine if an unhandled exception has occurred:
+    - ExceptionThrown: Record the originating FunctionID and ObjectID of the exception that is thrown.
+    - ExceptionSearchCatcherFound: Record that the exception will be handled in a catching FunctionID.
+    - ExceptionUnwindFunctionEnter: If the exception was not handled (ExceptionSearchCatcherFound was not invoked),
+      then the exception is unhandled at this point. Do some extra bookkeeping in the case when the exception is
+      handled so that the exception is cleared once frame with the corresponding catching FunctionID is encountered.
+
+    Additionally, use GC callbacks, such as MovedReferences#, to update the exception ObjectID when compacting occurs.
+
+    Current pitfalls:
+    - Algorithm doesn't account for exceptions thrown within exception filters. These will be seen as a new set of
+      callback invocations starting with throwing, searching, and unwinding, however, the callstack does not unwind
+      out of the exception filter and the original exception processing is resumed.
+    - Algorithm doesn't account for exceptions thrown within finally blocks. Presumably, these will be seen as new set
+      of callback invocations starting with throwing, searching, and unwinding. The original exception processing is
+      superceded by the new set of callbacks for the new exception.
+*/
+
+
+ExceptionTracker::ExceptionTracker(
+    const shared_ptr<ILogger>& pLogger,
+    const shared_ptr<ThreadDataManager> pThreadDataManager,
+    ICorProfilerInfo2* pCorProfilerInfo)
+{
+    m_pCorProfilerInfo = pCorProfilerInfo;
+    m_pLogger = pLogger;
+    m_pThreadDataManager = pThreadDataManager;
+}
+
+void ExceptionTracker::AddProfilerEventMask(DWORD& eventsLow)
+{
+    eventsLow |= COR_PRF_MONITOR::COR_PRF_ENABLE_STACK_SNAPSHOT;
+    eventsLow |= COR_PRF_MONITOR::COR_PRF_MONITOR_GC;
```

Comment:
I am very concerned about GC Tracking performance. It has impact on the GC mode and other issues such as https://github.com/dotnet/runtime/issues/53326.

Reply:
I understand the concerns here.

However, we need to be able to keep a valid ObjectID for the thrown exceptions. The only callback that runs cooperatively is ExceptionThrown, which means that compacting GC can occur during the other callbacks (they are preemptive). I have an easy repro that demonstrates a thrown exception is moved as part of a compacting GC while it is unwinding a non-async call stack.

We also need to track the ObjectID outside of ExceptionThrown so that in a future change I can flag the first occurrence of a thrown exception as interesting and likely want to ignore subsequent rethrows; these rethrows have the same ObjectID as the original throwing of the exception.

I'm completely fine with using some other mechanism for tracking this information so that we can avoid the GC callbacks (e.g. maybe we can pin the exception instance from the profiler callback), but am currently not sure what will work.

Reply:
I've remove the storage of the exception ObjectID for now since it is not strictly necessary do this in order to detect that an unhandled exception has occurred.